### PR TITLE
Shorten fc lock

### PR DESCRIPTION
**Pull request by potuz**
This PR removes the lock held during the whole of block processing. 

On the good side: 
- It does not lock forkchoice during consensus and execution validation which is the biggest part of the process. 
- It cleans up a little bit the order or operations on `onBlock`. 

On the bad side: 
- Code writers of the blockchain package need to remember to acquire locks appropriately. In particular, locking a function that calls the `chain-info.go` functions (like this PR wiill do) will result in a deadlock.

### CODE REVIEW

File: beacon-chain/blockchain/process_block.go [hunk 1]
```diff
@@ -134,6 +135,9 @@ func (s *Service) onBlock(ctx context.Context, signed interfaces.ReadOnlySignedB
 	}
 	isValidPayload, err := s.notifyNewPayload(ctx, postStateVersion, postStateHeader, signed)
 	if err != nil {
+		if IsInvalidBlock(err) && InvalidBlockLVH(err) != [32]byte{} {
```
Comment (rkapka) [inline beacon-chain/blockchain/process_block.go:138]: Why don't we want to report when `LVH == [32]byte{}`? Isn't it still an invalid block?

File: beacon-chain/blockchain/process_block.go [hunk 2]
```diff
@@ -145,19 +149,38 @@ func (s *Service) onBlock(ctx context.Context, signed interfaces.ReadOnlySignedB
...
 	}
+	start := time.Now()
+	headRoot, err := s.cfg.ForkChoiceStore.Head(ctx)
+	if err != nil {
+		log.WithError(err).Warn("Could not update head")
+	}
+	newBlockHeadElapsedTime.Observe(float64(time.Since(start).Milliseconds()))
+
+	if err := s.handleBlockAttestations(ctx, signed.Block(), postState); err != nil {
+		return errors.Wrap(err, "could not handle block's attestations")
```
Comment (rkapka) [inline beacon-chain/blockchain/process_block.go:174]: `s.cfg.ForkChoiceStore.Unlock()` missing

File: beacon-chain/blockchain/process_block.go [hunk 3]
```diff
@@ -185,13 +208,8 @@ func (s *Service) onBlock(ctx context.Context, signed interfaces.ReadOnlySignedB
 			}
 		}()
 	}
-	justified := s.cfg.ForkChoiceStore.JustifiedCheckpoint()
-	start := time.Now()
-	headRoot, err := s.cfg.ForkChoiceStore.Head(ctx)
-	if err != nil {
-		log.WithError(err).Warn("Could not update head")
-	}
 	if blockRoot != headRoot {
```
Comment (rkapka) [inline beacon-chain/blockchain/process_block.go:211]: I think you have to fetch head from forkchoice here. You moved `forkchoiceUpdateWithExecution` above, meaning that the head can change (on a side note, this looks like a possibly unwanted change in logic). Actually I see that you ignore the boolean return value from `forkchoiceUpdateWithExecution` - you can use it to determine whether the head has changed.

File: beacon-chain/blockchain/execution_engine.go [hunk 4]
```diff
@@ -220,35 +220,37 @@ func (s *Service) notifyNewPayload(ctx context.Context, postStateVersion int,
...
+		lvh := bytesutil.ToBytes32(lastValidHash)
 		return false, invalidBlock{
-			invalidAncestorRoots: invalidRoots,
-			error:                ErrInvalidPayload,
+			error:         ErrInvalidPayload,
+			lastValidHash: lvh,
 		}
-	case execution.ErrInvalidBlockHashPayloadStatus:
-		newPayloadInvalidNodeCount.Inc()
-		return false, ErrInvalidBlockHashPayloadStatus
```
[the remark this answers is not in the document]
Comment (potuz) [reply beacon-chain/blockchain/execution_engine.go:246]: yeah thanks, my first design was to check that the returned error was this one, but went with non-trivial LVH in the end.

File: beacon-chain/blockchain/process_block.go (hunk 1 above)
Comment (potuz) [reply beacon-chain/blockchain/process_block.go:138]: we do return with an error, but if the LVH is not available there's nothing we can do in forkchoice: the invalid block never made it to forkchoice  yet and there's no info to remove parents then

File: beacon-chain/blockchain/process_block.go (hunk 2 above)
Comment (potuz) [reply beacon-chain/blockchain/process_block.go:174]: nice catch, this was supposed to be changed to a log only

File: beacon-chain/blockchain/execution_engine.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (potuz) [reply beacon-chain/blockchain/execution_engine.go:246]: oh this is a different one! thanks nice catch

File: beacon-chain/blockchain/process_block.go [hunk 5]
```diff
@@ -149,19 +149,38 @@ func (s *Service) onBlock(ctx context.Context, signed interfaces.ReadOnlySignedB
...
+	start := time.Now()
+	headRoot, err := s.cfg.ForkChoiceStore.Head(ctx)
+	if err != nil {
+		log.WithError(err).Warn("Could not update head")
+	}
+	newBlockHeadElapsedTime.Observe(float64(time.Since(start).Milliseconds()))
+
+	if err := s.handleBlockAttestations(ctx, signed.Block(), postState); err != nil {
+		log.WithError(err).Warn("could not handle block's attestations")
+	}
```
Comment (terencechain) [inline beacon-chain/blockchain/process_block.go:175]: It looks like you changed the ordering. `handleBlockAttestations` should be before `Head`. Do we not want to count the attestations in the block?

**Review by terencechain — CHANGES_REQUESTED** [review]
We are missing locks here

`getBlockPreState` -> `verifyBlkFinalizedSlot` ->  `s.cfg.ForkChoiceStore.FinalizedCheckpoint()`

`reportInvalidBlock` ->  `s.cfg.ForkChoiceStore.SetOptimisticToInvalid`

**Review by rauljordan — APPROVED** [review]

**Review by terencechain — APPROVED** [review]

**Review by terencechain — DISMISSED** [review]

File: beacon-chain/blockchain/process_block.go [hunk 6]
```diff
@@ -149,19 +149,38 @@ func (s *Service) onBlock(ctx context.Context, signed interfaces.ReadOnlySignedB
...
 	}
+	start := time.Now()
+	if err := s.handleBlockAttestations(ctx, signed.Block(), postState); err != nil {
+		log.WithError(err).Warn("could not handle block's attestations")
+	}
+	headRoot, err := s.cfg.ForkChoiceStore.Head(ctx)
+	if err != nil {
+		log.WithError(err).Warn("Could not update head")
+	}
+	newBlockHeadElapsedTime.Observe(float64(time.Since(start).Milliseconds()))
```
Comment (nisdas) [inline beacon-chain/blockchain/process_block.go:174]: shouldn't this metric be at the end of this code block ?

Comment (nisdas) [reply beacon-chain/blockchain/process_block.go:174]: nvm ignore, we are just following the same order as before

**Review by nisdas — DISMISSED** [review]

**Comment by potuz** [comment]
Closing in favour of #12589

**Pull request closed without merging**